freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/PasswordService.java

```diff
@@ -41,48 +54,103 @@
     @Inject
     private GrpcUmsClient umsClient;
 
-    public SetPasswordResponse setPassword(String accountId, String userCrn, String password, Set<String> envs) {
-        String userId = getUserIdFromUserCrn(userCrn);
+    @Inject
+    private AsyncTaskExecutor asyncTaskExecutor;
+
+    // TODO replace cache with database tables
+    private Cache<String, SetPasswordResponse> passwordOperationStatusCache = CacheBuilder.newBuilder()
+            .maximumSize(MAX_PASSWORD_OPERATION_CACHE_SIZE)
+            .expireAfterAccess(PASSWORD_OPERATION_EXPIRATION)
+            .build();
 
+    public SetPasswordResponse setPassword(String accountId, String userCrn, String password, Set<String> envs) {
         LOGGER.debug("setting password for user {} in account {}", userCrn, accountId);
 
-        List<Stack> stacks = stackService.getAllByAccountId(accountId);
+        String operationId = UUID.randomUUID().toString();
+        long startTime = System.currentTimeMillis();
+
+        List<Stack> stacks;
+        List<Stack> allStacks = stackService.getAllByAccountId(accountId);
+        // Filter based on provided envs
+        if (envs != null && !envs.isEmpty()) {
+            LOGGER.debug("Environment filter provided for environments :{}", envs);
+            stacks = allStacks.stream()
+                    .filter(stack -> envs.contains(stack.getEnvironmentCrn()))
+                    .collect(Collectors.toList());
+            if (LOGGER.isDebugEnabled()) {
+                stacks.forEach(stack -> LOGGER.debug("Found matching stack for environment {}", stack.getEnvironmentCrn()));
+            }
+        } else {
+            stacks = allStacks;
+        }
 
         if (stacks.isEmpty()) {
-            throw new IllegalArgumentException("No stacks found for accountId " + accountId);
+            LOGGER.warn("No stacks found for accountId {}", accountId);
+            throw new NotFoundException("No stacks found for accountId " + accountId);
+        }
+
+        Future<List<SetPasswordRequest>> requests = asyncTaskExecutor.submit(() -> triggerSetPasswords(accountId, userCrn, password, stacks));
+
+        SetPasswordResponse response = SetPasswordResponse.running(operationId, startTime);
+        passwordOperationStatusCache.put(operationId, response);
+
+        asyncTaskExecutor.submit(() -> waitSetPasswords(operationId, accountId, userCrn, requests));
```

Comment:
do we really need 2 thread here? the 2nd is blocked on waiting for the 1st one for most of the time doing nothing